### PR TITLE
Transaction memos

### PR DESCRIPTION
**Pull request by eb-concordium**
## Purpose

To add user-defined metadata to inter-account transfers. This will enable senders to attach free-form data to transfers that will enable them to convey additional information to the receiver. 

Closes https://github.com/Concordium/concordium-base/issues/63 and parts of https://github.com/Concordium/concordium-node/issues/115. 

## Changes

So far the following changes has been done:
- Three new transaction types that are extensions of Transfer, EncryptedTransfer and TransferWithSchedule have been added.
- Event type has been extended with an additional Metadata event which is generated by the three new transactions. 
- The decodePayload function that takes a serialized payload is parametrized by a protocol version
- The mobile wallet library has been updated to support the transfers with memo.

## Checklist

- [ ] My code follows the style of this project.
- [ ] The code compiles without warnings.
- [ ] I have performed a self-review of the changes.
- [ ] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [ ] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: haskell-src/Concordium/Types/ProtocolVersion.hs [hunk 1]
```diff
@@ -18,25 +18,46 @@
...
 instance Serialize ProtocolVersion where
     put P1 = putWord64be 1
+    put P2 = putWord64be 2
     get =
         getWord64be >>= \case
             1 -> return P1
+            2 -> return P2
             v -> fail $ "Unknown protocol version: " ++ show v
 
+instance ToJSON ProtocolVersion where
```
[the remark this answers is not in the document]
Comment (abizjak) [reply haskell-src/Concordium/Types/ProtocolVersion.hs:52]: Done.

File: haskell-src/Concordium/Types.hs [hunk 2]
```diff
@@ -598,6 +601,44 @@ instance S.Serialize Nonce where
...
+  put (Memo bss) = do
+    S.putWord16be . fromIntegral . BSS.length $ bss
+    S.putShortByteString bss
+
+  get = do
+    l <- fromIntegral <$> S.getWord16be
+    unless (l <= maxMemoSize) $ fail "Data too long"
+    Memo <$> S.getShortByteString l
+
+instance AE.FromJSON Memo where
```
[the remark this answers is not in the document]
Comment (abizjak) [reply haskell-src/Concordium/Types.hs:639]: Added comment as to why the situation is the way it is.

File: haskell-src/Concordium/Types/Queries.hs [hunk 3]
```diff
@@ -75,7 +75,18 @@ data ConsensusStatus = ConsensusStatus
...
+      csProtocolVersion :: !ProtocolVersion,
+      -- |The number of chain restarts via a protocol update. An effected
+      -- protocol update instruction might not change the protocol version
+      -- specified in the previous field, but it always increments the genesis
+      -- index.
+      csGenesisIndex :: !GenesisIndex,
+      -- |Block hash of the current era, i.e., since the last protocol update. Initially this is equal to 'csGenesisBlock'.
+      csCurrentEraGenesisBlock :: !BlockHash,
+      -- |Time when the current era started.
+      csCurrentEraGenesisTime  :: !UTCTime
```
[the remark this answers is not in the document]
Comment (abizjak) [reply haskell-src/Concordium/Types/Queries.hs:89]: I don't think this is really necessary. While in principle it would be nice, doing this would mean manually writing out the FromJSON instance which is time consuming and will be a continual maintenance burden.

File: haskell-src/Concordium/Genesis/Data/P2.hs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+import Data.Serialize
+
+import Concordium.Common.Version
+import qualified Concordium.Crypto.SHA256 as Hash
+import Concordium.Genesis.Data.Base
+import Concordium.Types
+
+-- |Genesis data for the P2 protocol version.
+newtype GenesisDataP2
+    = GenesisDataP2 { unGenesisDataP2 :: RegenesisData }
```
[the remark this answers is not in the document]
Comment (abizjak) [reply haskell-src/Concordium/Genesis/Data/P2.hs]: I have added this now by essentially copying the P1 one.

I thought that we should instead have just serialized state instead of the genesis parameters, but that causes more problems since the rest of the types/infrastructure is not ready, and I do not have time to spend refactoring things for no clear gain at present.

If this ever becomes an issue it can be revisited in the future.

[the remark this answers is not in the document]
Comment (abizjak) [reply haskell-src/Concordium/Genesis/Data/P2.hs]: I'll push the update to the genesis tool later tonight.

[the remark this answers is not in the document]
Comment (abizjak) [reply haskell-src/Concordium/Genesis/Data/P2.hs]: Done.

File: rust-src/crypto_common/src/types.rs [hunk 5]
```diff
@@ -212,6 +212,59 @@ impl<'de> SerdeDeserialize<'de> for Amount {
...
+        (self.memo.len() as u16).serial(out);
+        out.write_all(&self.memo)
+            .expect("Writing to buffer should succeed.");
+    }
+}
+
+impl Deserial for Memo {
+    fn deserial<R: ReadBytesExt>(source: &mut R) -> ParseResult<Self> {
+        let len: u16 = source.get()?;
+        if len as usize > MAX_MEMO_SIZE {
```
Comment (abizjak) [inline rust-src/crypto_common/src/types.rs:233]: Suggest to shorten this to
```rust
anyhow::ensure!(len as usize <= MAX_MEMO_SIZE, "Memo size of {} is too big. Maximum size is {}.", len, MAX_MEMO_SIZE)
```

File: haskell-bins/genesis/README.md [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+After these files are available `initial-accounts.json` may be generated using
+another [genesis-tool](../../rust-bins/src/bin/genesis_tool.rs).
+
+The update authorizations contain public keys for performing chain updates. For
+testing purposes they may be generated using the
+[generate-update-keys](../generate-update-keys/Main.hs) tool.
+
+### Supported protocol versions
+
+- If `--gdver=3` then the tool will output initial genesis for protocol version 1.
```
Comment (td202) [inline haskell-bins/genesis/README.md:94]:
```suggestion
- If `--gdver=3` then the tool will output initial genesis for protocol version 1.
  (This is the default if no version is specified.)
```

**Review by td202 — APPROVED** [review]

File: haskell-src/Concordium/Types/Queries.hs (hunk 3 above)
[the remark this answers is not in the document]
Comment (abizjak) [reply haskell-src/Concordium/Types/Queries.hs:89]: As discussed today at the meeting, we do not aim for backwards compatibility in this way.

**Review by abizjak — APPROVED** [review]

**Comment by abizjak** [comment]
Looks good @eb-concordium . Please merge and then update the submodule reference in node and concordium-client.
